### ISSUE TITLE
change job priority of ARM jobs

### ISSUE THREAD

**Issue by dirk-thomas** [ros-infrastructure/ros_buildfarm_config#19]
Since ARM jobs take a lot of time they tend to block other jobs from running for days. I would suggest we give them a lower priority then all other job types. Basically only let them run when the build farm has no other jobs to run.

**Comment by tfoote** [comment]
It's something to consider. However I worry that if we deprioritize them too much it will be much harder for us to complete a rebuild.

**Comment by dirk-thomas** [comment]
Can you estimate how long a full rebuild of e.g. Indigo takes on ARM compared to non-ARM? I would expect many days and think that blocking e.g. doc jobs for that long is not a good idea.

**Comment by tfoote** [comment]
It highly depends on the size of the farm. The arm rebuilds tend to take 5-10x the time of the x86 builds. If the farm is small a full rebuild can take a very long time(weeks) if it doesn't scale up. If it scales up, the arm builds are now parallelizable enough to saturate the farm and the doc jobs will flow through the unused executors. I think that we should be able to detect the multi-day size buildup and scale the farm. 

I also want to look into setting up autoscaling the farm. There are EC2 jenkins plugins for that which could resolve this too.

**Comment by dirk-thomas** [comment]
Why not give the ARM jobs a lower priority then? Then they don't hold up any of the other builds. And if the backlog gets to big the number of executors need to be increased anyway. So we get a clear benefit but no drawback.

**Comment by tfoote** [comment]
I don't agree that the doc jobs should be higher than the arm debbuilds but lower than the x86. The x86 jobs will starve the doc jobs for several days too if the farm does not scale up. 

We can make the doc jobs higher priority than both, but I don't think that putting them between the x86 and the arm jobs makes sense.

**Comment by dirk-thomas** [comment]
I think giving a maintainer of a released package fast response is important. Therefore a source job (as well as a binary job) should be run rather sooner then later. doc jobs could significantly delay that.

**Comment by wjwwood** [comment]
Well ideally you'd have a weighted queue, where jobs start with a particular value which is increased the longer it is in the queue. Obviously that probably isn't an option, though it might be worth looking for a plugin that can do that.

Otherwise, what about giving the doc jobs and the arm jobs equal priority? Then it would alternate through them as arm and doc jobs are scheduled.

Another idea is to leave the priorities, but have a few core slaves which are doc job only or arm job only, but the vast majority will build any job according to the priority queue. That way you never get complete starvation.

If all of those ideas don't work then I'd say that putting the arm jobs behind the doc jobs would be a reasonable optimization. Though I do agree with @tfoote that the distinction is pretty arbitrary, e.g. why not put the 32-bit jobs behind the doc jobs too. It seems to me that only one binary job type needs to be ahead of other jobs to give the developers a timely answer to whether or not their binary package failed. There are cases where a package will build on one configuration but not another, but if you argue for that then I'd argue that arm is just as important as 32-bit or some other flavor of Ubuntu that is less commonly used.

On balance I think my favorite solution is to keep arm jobs at the same priority as x86/x86_64 and make a few doc job only slaves.

**Comment by dirk-thomas** [comment]
- We don't have different priority options right now.
- The same priority would still result in a massive delay of the `doc` jobs when triggered after the `arm` jobs.
- I don't think we can effort having dedicated machines for specific job types.
- The idea to put `doc` between `arm` and `amd64` / `i686` is not arbitrary, it is based on the average runtime of these jobs. We want fast feedback on releases (for as many platforms and arches as possible) but we don't want `doc` jobs to wait a week for arm jobs to finish.

**Comment by wjwwood** [comment]
Why can't we have a few slaves on a few machines either be doc only or everything minus arm?

**Comment by tfoote** [comment]
We have worked hard to make the machines homogeneous such that any job can run anywhere. And we scale the number of machines from 1 to ~10 depending on the load. We don't have the ability to easily force a distribution of labels that scales with size. And it will only lead to machines not being used when they could be. That's why we use the priorities such that the prioritized jobs go first. There's a level that if things get backed up there is always something that will have to wait. We set the priorities such that the things with lowest priority wait the longest. As nice as it is for the documentation to be up to date instantly I think that the binaries are actually more important. We only run the doc jobs every 3 days. If they have an additional wait of a few days it's not a significant change. We're not promising doc jobs to run with the same speed as a devel job or a PR job. The are a background task that you don't sit and wait for.  That's why I think that they should stay at the back. 

If we don't want the doc jobs to wait a week we should scale up the farm, because any binary jobs not finishing also blocks the sync. And although the arm jobs take a long time to finish, it's only during the early phases when they have enough parallelization which can starve the farm when spun up. At the end they have a very long tail and the doc jobs will likely finish before the arm debbuilds as they can filter through the unused executors near the end.
